corehq/apps/userreports/pillow.py

```diff
@@ -208,6 +210,120 @@ def _save_doc_to_table(self, domain, table, doc, eval_context):
             # remove it until the next bootstrap call
             self.table_adapters_by_domain[domain].remove(table)
 
+    def _get_ucr_values(self, domain, table, doc, eval_context):
+        try:
+            return table.get_all_values(doc, eval_context)
+        except Exception as e:
+            try:
+                table.handle_exception(doc, e)
+            except UserReportsWarning:
+                # remove it until the next bootstrap call
+                self.table_adapters_by_domain[domain].remove(table)
+
+    def process_changes_chunk(self, pillow_instance, changes):
+        """
+        Update UCR tables in bulk by breaking up changes per domain per UCR table.
+            If an exception is raised in bulk operations of a set of changes,
+            those changes are returned to pillow for serial reprocessing.
+        """
+        self.bootstrap_if_needed()
+        # break up changes by domain
+        changes_by_domain = defaultdict(list)
+        for change in changes:
+            # skip if no domain or no UCR tables in the domain
+            if change.metadata.domain and change.metadata.domain in self.table_adapters_by_domain:
+                changes_by_domain[change.metadata.domain].append(change)
+
+        retry_changes = set()
+        for domain, changes_chunk in six.iteritems(changes_by_domain):
+            failed = self._process_chunk_for_domain(domain, changes_chunk)
+            retry_changes.update(failed)
+
+        return retry_changes
+
+    def _process_chunk_for_domain(self, domain, changes_chunk):
+        all_adapters = list(self.table_adapters_by_domain[domain])
+        to_delete_by_adapter = defaultdict(list)
+        rows_to_save_by_adapter = defaultdict(list)
+        async_configs_by_doc_id = defaultdict(list)
+        to_update = {change for change in changes_chunk if not change.deleted}
+        retry_changes, docs = self.get_docs_for_changes(to_update, domain)
+
+        for doc in docs:
+            # make a shallow copy so that adapters can be removed
+            adapters = list(self.table_adapters_by_domain[domain])
```

Comment:
did you mean to make the copy inside the docs for loop? That seems to defeat the purpose of removing the table in the first place.